### PR TITLE
Rewrite the `EventListeners` Page

### PR DESCRIPTION
**Pull request by sergetron**
# Changes

Rewrites the `EventListener` page for clarity and flow, breaks out rewritten `Interceptor` documentation into the already published `interceptors.md` page, and merges rewritten content from the `exposing-eventlisteners.md` page.

# Submitter Checklist

These are the criteria that every PR should meet, please check them off as you
review them:

- [ ] Includes [tests](https://github.com/tektoncd/community/blob/master/standards.md#tests) (if functionality changed/added)
- [X] Includes [docs](https://github.com/tektoncd/community/blob/master/standards.md#docs) (if user facing)
- [ ] Commit messages follow [commit message best practices](https://github.com/tektoncd/community/blob/master/standards.md#commits)
- [ ] Release notes block has been filled in or deleted (only if no user facing changes)

_See [the contribution guide](https://github.com/tektoncd/triggers/blob/master/CONTRIBUTING.md) for more details._

# Release Notes
```release-note
NONE
```

### CODE REVIEW

File: docs/eventlisteners.md [hunk 1]
```diff
@@ -297,94 +306,58 @@ spec:
...
-ServiceAccountName
-NodeSelector
-Tolerations
-Volumes
-Containers
-- Resources
-- VolumeMounts
-- Env
-```
+## Specifying target namespaces
```
Comment (dibyom) [inline docs/eventlisteners.md:309]: Do you think its clearer if we say something like Specifying Triggers in target namespaces? (And for the label selector below, - Specifying Triggers using labels?)

File: docs/eventlisteners.md [hunk 2]
```diff
@@ -430,618 +384,215 @@ metadata:
     service.beta.kubernetes.io/aws-load-balancer-backend-protocol: https
 ```
 
-**Note**: If there are any annotations attached to the service or the deployment, they will get overwritten 
-by the annotations available in the eventlistener.
-
-## Interceptors
```
Comment (dibyom) [inline docs/eventlisteners.md:436]: Should we at least have a link to the Interceptors page here ?

File: docs/eventlisteners.md [hunk 3]
```diff
@@ -430,618 +384,215 @@ metadata:
...
 ```
-- `eventListener` - Refers to the EventListener Name.
-- `namespace` - Refers to the namespace of the EventListener
-- `eventID` - Refers to the uniqueID that gets assigned to each incoming request
+- `eventListener` - name of the target EventListener
+- `namespace` - namespace of the target EventListener
+- `eventID` - UID assigned to this event request
 
-## How does the EventListener work?
+An `EventListener` times out if it cannot process an event request within a timeout specified in [controller.yaml](../config/controller.yaml). The timeouts are as follows:
```
Comment (dibyom) [inline docs/eventlisteners.md:401]: This should probably have its own heading? e.g. specifying EventListener Timeouts?

File: docs/eventlisteners.md [hunk 4]
```diff
@@ -430,618 +384,215 @@ metadata:
...
+- `REASON` - reason for the value displayed in the `AVAILABLE` column
+- `READY` - readiness state of the Kubernetes Custom Resource object specified in the `EventListener`
+- `REASON` - reason for the value displayed in the `READY` column
 
-Triggers now support both `HTTP` and `HTTPS` connection by adding few configuration to eventlistener.
+**Note:** The status messaging described above is being refactored. For more information, see [Issue 932](https://github.com/tektoncd/triggers/issues/932).
 
-To setup TLS connection add two set of reserved environment variables `TLS_CERT` and `TLS_KEY` using `secretKeyRef` env type 
-where we need to specify the `secret` which contains `cert` and `key` files. See the full [example](../examples/eventlistener-tls-connection/README.md) for more details.
+## Using `EventListener` logging
```
Comment (dibyom) [inline docs/eventlisteners.md:437]: Specifying EventListener logging options?

Comment (dibyom) [reply docs/eventlisteners.md:437]: Or Configuring Logging for EventListener?

Comment (sergetron) [reply docs/eventlisteners.md:437]: Reworded to "Configuring logging for `EventListeners`".

File: docs/eventlisteners.md [hunk 5]
```diff
@@ -1,111 +1,95 @@
...
+An `EventListener` is a Kubernetes object that listens for events at a specified port on your Kubernetes cluster. 
+It exposes an addressable sink that receives incoming event and specifies one or more [`Triggers`](./triggers.md).
+The sink is a Kubernetes service running the sink logic inside a dedicated Pod.
+
+Each `Trigger`, in turn, allows you to specify one or more [`TriggerBindings`](./triggerbindings.md) that allow
+you to extract fields and their values from event payloads, and one or more [`TriggerTemplates`](./triggertemplates.md)
+that receive field values from the corresponding `TriggerBindings` and allow Tekton Triggers to instantiate resources
+such as `TaskRuns` and `PipelineRuns` with that data. 
+
+If you need to modify the event payload data before passing it to a `TriggerBinding`, you can optionally specify one
```
[the remark this answers is not in the document]
Comment (sergetron) [reply docs/eventlisteners.md:18]: Added validation and filtering to the sentence.

File: docs/eventlisteners.md [hunk 6]
```diff
@@ -1,111 +1,95 @@
...
+  - [Specifying a `kubernetesResource` object](#specifying-a-kubernetes-resource-object)
+  - [Specifying a `CustomResource` object](#specifying-a-custom-resource-object)
+- [Specifying target namespaces](#specifying-target-namespaces)
+- [Labels in `EventListeners`](#labels-in-eventlisteners)
+  - [Specifying target labels](#specifying-target-labels)
+- [Annotations in `EventListeners`](#annotations-in-eventlisteners)
+- [Understanding `EventListener` response](#understanding-eventlistener-response)
+- [TLS HTTPS support in `EventListeners`](#tls-https-support-in-eventlisteners)
+- [Obtaining the status of deployed `EventListeners`](#obtaining-the-status-of-deployed-eventlisteners)
+- [Using `EventListener` logging](#using-eventlistener-logging)
```
[the remark this answers is not in the document]
Comment (sergetron) [reply docs/eventlisteners.md:36]: Changed to "Configuring logging for `EventListeners`'.

File: docs/eventlisteners.md [hunk 7]
```diff
@@ -1,111 +1,95 @@
...
-  - [`metadata`][kubernetes-overview] - Specifies data to uniquely identify the
-    `EventListener` resource object, for example a `name`.
-  - [`spec`][kubernetes-overview] - Specifies the configuration information for
-    your EventListener resource object. In order for an EventListener to do
-    anything, the spec must include:
-    - [`triggers`](#triggers) - Specifies a list of Triggers to run
-    - [`serviceAccountName`](#serviceAccountName) - Specifies the ServiceAccount
-      that the EventListener uses to create resources
+  - [`apiVersion`][kubernetes-overview] - specifies the target API version, for example `triggers.tekton.dev/v1alpha1`
+  - [`kind`][kubernetes-overview] - specifies that thisKubernetes resource is an `EventListener` object
```
[the remark this answers is not in the document]
Comment (sergetron) [reply docs/eventlisteners.md:53]: Fixed missing space.

File: docs/eventlisteners.md [hunk 8]
```diff
@@ -1,111 +1,95 @@
...
+  - [`metadata`][kubernetes-overview] - specifies data that uniquely identifies this `EventListener` object, for example a `name`
+  - [`spec`][kubernetes-overview] - specifies the configuration of your `EventListener`:
+    - [`triggers`](#triggers) - Specifies a list of `Triggers` to execute upon event detection
```
[the remark this answers is not in the document]
Comment (sergetron) [reply docs/eventlisteners.md:56]: Moved the `triggers` field to the "Optional" list.

File: docs/eventlisteners.md [hunk 9]
```diff
@@ -1,111 +1,95 @@
...
-    for your EventListener pod
-  - [`resources`](#resources) - Specifies the Kubernetes/Custom Resource shape for the EventListener sink
-    for your EventListener pod
-  - [`namespaceSelector`](#namespaceSelector) - Specifies the namespaces where
-    EventListener can fetch triggers from and create Tekton resources.
+  - [`serviceType`](#serviceType) - specifies the type of service you want the `EventListener` to use to listen for events
+  - [`replicas`](#specifying-a-kubernetes-resource-object) - specifies the number of `EventListener` pods to create (only for `kubernetesResource` objects)
+  - [`podTemplate`](#podTemplate) - specifies the `PodTemplate` for your `EventListener` pod
+  - [`resources`](#resources) - specifies the resources that will be available to the event listening service
+  - [`namespaceSelector`](#namespaceSelector) - specifies the namespace for the `EventListener`; this is where the `EventListener` looks for the
```
[the remark this answers is not in the document]
Comment (sergetron) [reply docs/eventlisteners.md:63]: Added the `labelSelector` field and its description.

File: docs/eventlisteners.md [hunk 10]
```diff
@@ -1,111 +1,95 @@
...
+- A Kubernetes `ClusterRole` that permits read access to `ClusterTriggerBindings` objects
+- Permissions to create the Tekton resources specified in the associated `TriggerTemplate`, as shown in the following [example](../examples/rbac.yaml)
+- If you're using `namespaceSelectors` in your `EventListener`, a `ClusterRole` that permits read access to all `Trigger` objects on the cluster.ources across the cluster.
 
-### Triggers
+## Specifying `Triggers`
 
-The `triggers` field is required. Each EventListener can consist of one or more
-`triggers`. A Trigger consists of:
+You must specify one or more `Triggers` that define the actions to take when the `EventListener` detects a qualifying event. You can specify *either* a reference to an
```
[the remark this answers is not in the document]
Comment (sergetron) [reply docs/eventlisteners.md:82]: Reworded to indicate specifying `Triggers` is optional.

File: docs/eventlisteners.md [hunk 11]
```diff
@@ -1,111 +1,95 @@
...
+external `Trigger` object *or* reference/define the `TriggerBindings`, `TriggerTemplates`, and `Interceptors` in the `Trigger` definition. A `Trigger` definition 
+specifies the following fields:
 
-- `name` - (Optional) a valid
-  [Kubernetes name](https://kubernetes.io/docs/concepts/overview/working-with-objects/labels/#syntax-and-character-set)
-- [`interceptors`](#interceptors) - (Optional) list of interceptors to use
-- `bindings` - (Optional) A list of bindings to use. Can either be a reference to existing `TriggerBinding` resources or embedded name/value pairs.
-- `template` - (Optional) Either a reference to a TriggerTemplate object or an embedded TriggerTemplate spec.
-- `triggerRef` - (Optional) Reference to the [`Trigger`](./triggers.md).
+- `name` - (optional) a valid[Kubernetes name](https://kubernetes.io/docs/concepts/overview/working-with-objects/labels/#syntax-and-character-set) that uniquely identifies the `Trigger`
```
[the remark this answers is not in the document]
Comment (sergetron) [reply docs/eventlisteners.md:86]: Added missing space.

File: docs/eventlisteners.md [hunk 12]
```diff
@@ -216,25 +195,34 @@ spec:
       effect: NoSchedule
 ```
 
-### Resources
+## Specifying `Resources`
 
-The `resources` field is optional.
-Resource field helps to provide Kubernetes or custom resource information.
+You can optionally specify one or more Kubernetes `Resources` in your `EventListener` using the `resources` field. It accepts the following types of objects:
```
[the remark this answers is not in the document]
Comment (sergetron) [reply docs/eventlisteners.md:200]: Reworded as per suggestion.

File: docs/eventlisteners.md [hunk 13]
```diff
@@ -259,30 +247,51 @@ spec:
...
 
-##### Replicas
+**Note:** The `spec.replicas` field is now the `spec.resources.kubernetesResource.replicas` field.
 
-The `replicas` field is optional. By default, the number of replicas for the EventListener is 1.
-If you want to deploy more than one pod, you can specify the amount with the `replicas` field.
+### Specifying a `CustomResource` object
 
-**Note:** If the user sets the `replicas` field while creating/updating the eventlistener yaml then it won't respect the replicas value edited by the user manually on deployment or through any other mechanism (ex: HPA).
+You can specify a Kubernetes Custom Resource Definition (CRD) using the `CustomResource` field. This field has one sub-field, `runtime.RawExtension` that allows you to specify dynamic objects.
```
[the remark this answers is not in the document]
Comment (sergetron) [reply docs/eventlisteners.md:259]: Reworded as per suggestion.

File: docs/eventlisteners.md [hunk 14]
```diff
@@ -297,94 +306,58 @@ spec:
...
+
+| Name                              | Description                                                 |
+| --------------------------------- | ----------------------------------------------------------- |
+| triggers.tekton.dev/eventlistener | Name of the `EventListener` that instantiated the resource. |
+| triggers.tekton.dev/trigger       | Name of the `Trigger` that instantiated the resource.       |
+| triggers.tekton.dev/eventid       | UID of the incoming event.                                  |
+
+**Note:** Because they're used as labels, `EventListener` and `Trigger` names must conform to the [Kubernetes syntax and character set requirements](https://kubernetes.io/docs/concepts/overview/working-with-objects/labels/#syntax-and-character-set).
+
+### Specifying target labels
```
[the remark this answers is not in the document]
Comment (sergetron) [reply docs/eventlisteners.md:344]: Fixed section nesting.

File: docs/eventlisteners.md (hunk 2 above)
Comment (sergetron) [reply docs/eventlisteners.md:436]: Good catch! Added a "Specifying `Interceptors`" section with a short description and a link to the `Interceptors` page.

File: docs/eventlisteners.md (hunk 3 above)
Comment (sergetron) [reply docs/eventlisteners.md:401]: Broke this out into its own section called "Specifying `EventListener` timeouts".

File: docs/eventlisteners.md [hunk 15]
```diff
@@ -259,30 +247,51 @@ spec:
...
-**Note:** If the user sets the `replicas` field while creating/updating the eventlistener yaml then it won't respect the replicas value edited by the user manually on deployment or through any other mechanism (ex: HPA).
+You can specify a Kubernetes Custom Resource Definition (CRD) using the `CustomResource` field. This field has one sub-field, `runtime.RawExtension` that allows you to specify dynamic objects.
 
-#### CustomResource
-A `CustomResource` object has one field that supports dynamic objects.
-* runtime.RawExtension
+The CRD must satisfy the [WithPod{}](https://github.com/knative/pkg/blob/master/apis/duck/v1/podspec_types.go#L41) contract. For example:
 
-Here we will use a [Knative Service](https://knative.dev/docs/) as an example to demonstrate usage of `CustomResource`
+Contract-mandated CRD structure:
```
[the remark this answers is not in the document]
Comment (sergetron) [reply docs/eventlisteners.md:263]: Made the contract info its own sub-section and moved `WithPod` after it.

File: docs/eventlisteners.md [hunk 16]
```diff
@@ -259,30 +247,51 @@ spec:
...
+Contract-mandated CRD structure: 
+
+```spec
+   spec:
+     template:
+       metadata:
+       spec:
+```
+
+Contract-mandated CRD status check:
```
[the remark this answers is not in the document]
Comment (sergetron) [reply docs/eventlisteners.md:272]: Fixed.

File: docs/eventlisteners.md (hunk 1 above)
Comment (sergetron) [reply docs/eventlisteners.md:309]: Changed both to "Constraining `EventListeners` to specific namespaces/labels" let me know if that works.

**Comment by sergetron** [comment]
Addressed all review comments, PTAL!

**Comment by dibyom** [comment]
Should we remove the exposing-eventlisteners.md page?

**Comment by sergetron** [comment]
> Should we remove the exposing-eventlisteners.md page?

Yes, let's delete it in a new PR after we publish this rewrite.

**Comment by jmcshane** [comment]
It is still referenced in the content you are merging in this PR, so we'd have to ensure the content was appropriately merged into the new page. I agree tho, new PR for handing that page.

**Comment by sergetron** [comment]
> It is still referenced in the content you are merging in this PR, so we'd have to ensure the content was appropriately merged into the new page. I agree tho, new PR for handing that page.

The old exposing-eventlisteners.md page is not linked or mentioned anywhere in the rewritten content.

**Comment by jmcshane** [comment]
@sergetron ah, shoot..i saw the link to #exposing-eventlisteners and thought it was an external link

**Comment by sergetron** [comment]
> @sergetron ah, shoot..i saw the link to #exposing-eventlisteners and thought it was an external link

Still a good catch - that link was broken, just fixed it. :)

**Comment by sergetron** [comment]
All comments addressed and broken links fixed. PTAL.

**Comment by sergetron** [comment]
If all looks good can we get approval @dibyom?

**Pull request merged**